File: java/sieve/src/main/java/Sieve.java
import java.util.List;
import java.util.ArrayList;

public final class Sieve {

  private Integer limit;
  private List<Integer> primes;

  public Sieve(Integer limit) {
    this.limit = limit;
    primes = findPrimes();
  }

  public List<Integer> getPrimes() {
    return primes;
  }

  private List<Integer> findPrimes() {
    List<Integer> primes = new ArrayList<Integer>();
    List<Integer> possible = findPossible();

    while (!possible.isEmpty()) {
      Integer curr = possible.remove(0);
      possible = removeMultiples(curr, possible);
      primes.add(curr);
    }

    return primes;
  }

  private List<Integer> findPossible() {
    List<Integer> possible = new ArrayList<Integer>();

    for (int i = 2; i <= limit; i++) {
      possible.add(i);
    }

    return possible;
  }

  private List<Integer> removeMultiples(Integer curr, List<Integer> possible) {
    List<Integer> multiples = new ArrayList<Integer>();
    Integer multiplier = 2;

    while (!possible.isEmpty() &&
            curr * multiplier <= possible.get(possible.size() - 1)) {
      multiples.add(curr * multiplier);
      multiplier += 1;
    }

    possible.removeAll(multiples);
    return possible;
  }

}

File: go/sieve/sieve.go
package sieve

func Sieve(limit int) (primes []int) {
	if limit < 2 {
		return []int{}
	}
	unprocessed := createRange(2, limit)
	for len(unprocessed) != 0 {
		prime := unprocessed[0]
		unprocessed = unprocessed[1:]

		unprocessed = removeMultiples(unprocessed, prime)
		primes = append(primes, prime)
	}
	return primes
}

func removeMultiples(slice []int, prime int) (filtered []int) {
	if len(slice) == 0 {
		return slice
	}
	multiplyer := 2
	multiple := multiplyer * prime
	for len(slice) != 0 && multiple <= slice[len(slice)-1] {
		slice = remove(slice, multiple)
		multiplyer += 1
		multiple = multiplyer * prime
	}
	return slice
}

func remove(slice []int, value int) (result []int) {
	for _, v := range slice {
		if v != value {
			result = append(result, v)
		}
	}
	return result
}

func createRange(lower int, upper int) (result []int) {
	for i := lower; i <= upper; i++ {
		result = append(result, i)
	}
	return result
}
